### ISSUE TITLE
OLMo-3-7B-Think: tokenizer mismatch causes degenerate generation

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/ATLAS#12]
## Summary

OLMo-3-7B-Think (AllenAI, 7B params, 3 BF16 shards) loads and runs inference on the A100, but produces **degenerate output** — repetitive tokens like `.Sign.Sign.Sign` or `Comey Comey Comey` instead of coherent text. The root cause is a **tokenizer mismatch**: ATLAS's `atlas-tokenize` crate does not correctly implement OLMo-3's `dolma_v1_5` byte-level BPE vocabulary.

### GPU logits are structurally valid
- Logit spread: **114.8** (healthy range >10)
- Max probability: **0.96%** (healthy — no single token dominates)
- Token diversity: **10/10 unique** in top-10

### But argmax decoding picks wrong tokens
```
Prompt: "The capital of France is"
Expected: " Paris" (or similar)
Actual: ".Sign.Sign.Sign.Sign..." (repeated)
```

### SmolLM2-1.7B works correctly with the same pipeline
- Uses standard GPT-NeoX tokenizer that `atlas-tokenize` handles
- Generates coherent text at 12.6 tok/s on A100
- Proves the inference pipeline (GPU kernels, KV cache, attention, RoPE) is correct

## Root Cause Analysis

OLMo-3-7B uses the `dolma_v1_5` tokenizer (GPT-NeoX variant, vocab size **100,278**):
- Byte-level BPE with specific merge rules
- Token-to-id mapping differs from standard GPT-NeoX
- `atlas-tokenize` currently hardcodes a simplified BPE that works for GPT-2/NeoX-style tokenizers up to ~50K vocab

The mismatch means:
1. **Prompt encoding**: Input text gets wrong token IDs → model sees garbage input
2. **Output decoding**: Argmax token IDs map to wrong strings → output looks degenerate
3. **The model itself is fine** — logit distributions are healthy, weights loaded correctly in BF16

### Option A: Full `tokenizer.json` parser (Recommended)
Parse HuggingFace `tokenizer.json` directly in `atlas-tokenize`:
- Read `model.vocab` and `model.merges` from the JSON
- Implement byte-level BPE with the exact merge priority from the file
- Handles any HF-compatible model automatically
- **Estimated effort**: 3-5 days
- **Covers**: OLMo-3, Llama-3, Mistral, Phi, Qwen, and all future HF models

### Option B: Sentencepiece `.model` file support
Some models ship sentencepiece `.model` files. Parse the protobuf directly.
- More complex, covers fewer models than Option A
- Only needed if targeting models that *only* ship sentencepiece

### Option C: External tokenizer binary
Shell out to Python `transformers.AutoTokenizer` or the `tokenizers` Rust crate.
- Breaks ATLAS's zero-dependency philosophy
- Last resort

### Recommendation
**Option A** is the clear winner. The `tokenizer.json` format is well-documented, JSON-native (ATLAS already has JSON parsing), and covers 95%+ of HuggingFace models. This would make ATLAS truly model-agnostic.

## Validation Test

The existing test `gpu_inference_olmo3_quality_sanity` in `atlas-model` already gates on:
1. Gate 1: Logit spread > 10 ✅ (passes today)
2. Gate 2: Max probability < 50% ✅ (passes today)
3. Gate 3: Token diversity (informational only)

**Add Gate 4**: Coherence check — verify decoded output contains expected substrings for known prompts. This will catch tokenizer mismatches that produce valid logits but wrong text.

## Impact

- **OLMo-3-7B-Think cannot be served via atlas-api** until tokenizer is fixed
- **HuggingFace upload blocked**: Cannot publish `openhubresearch/ATLAS-OLMo-3-7B-Think-v4` when the model generates garbage through ATLAS
- **BF16 inference pipeline is validated** — 19.9 tok/s on A100 confirmed working (the bottleneck is purely tokenizer)
- **SmolLM2-1.7B works** as interim workaround (currently deployed on GPU server)

## Sprint Placement

This should be **Sprint 1 of v5.0** (Issue #10), as it's the tokenizer component of the "HuggingFace-native model loading" track. The fix would also satisfy the `tokenizer.json` requirement for Llama-3, Mistral, and other models on the v5 roadmap.

## Environment
- ATLAS v4.0.3 (tag `v4.0.3`, 532 tests passing)
- GPU: GCP a2-highgpu-1g, A100-SXM4-40GB, CUDA 12.9
- Model: `allenai/OLMo-3-7B-Think` (3 BF16 shards, 14GB, vocab 100,278)
- Working model: `HuggingFaceTB/SmolLM2-1.7B` (1 shard, 3.4GB, vocab 49,152)

**Comment by web3guru888** [comment]
## ✅ Fixed in v4.0.4

### Commits
- `d417c48` fix(atlas-tokenize): GPT-4 regex pre-tokenization
- `140a013` test(atlas-model): add gpu_olmo3_tokenizer_e2e
- `29c6742` test(atlas-model): update e2e with chat template + multi-gate

### What Changed
Implemented the full GPT-4/OLMo-3/LLaMA-3 pre-tokenization regex as a **zero-dependency hand-coded scanner** in `atlas-tokenize`. The previous `pretokenize()` was a naive space-splitter that produced wrong token IDs.

7 regex alternation patterns implemented:
1. Contractions: `(?i:'s|'t|'re|'ve|'m|'ll|'d)`
2. Words: `[^\r\n\p{L}\p{N}]?\p{L}+`
3. Numbers: `\p{N}{1,3}`
4. Punctuation: ` ?[^\s\p{L}\p{N}]+[\r\n]*`
5. Newlines: `\s*[\r\n]+` (with backtracking)
6. Trailing whitespace: `\s+(?!\S)` (with backtracking)
7. Whitespace: `\s+`

Auto-detects pre-tokenizer type from `tokenizer.json` config.

### Verified on A100
```
encode("The capital of France is") → [791, 6864, 315, 9822, 374] ✅
encode("Hello, world!")           → [9906, 11, 1917, 0]          ✅
encode("1234 tokens")             → [4513, 19, 11460]             ✅
Round-trip decode                 → perfect                       ✅
Chat template encoding            → 15 tokens, specials correct   ✅
GPU generation (30 tok, T=0.8)    → 30/30 unique, no repetition   ✅
```

### Tests
539 passed (532 → 539), including:
- 7 new pre-tokenizer unit tests
- 2 integration tests (OLMo-3 + SmolLM2 reference encodings)
- 1 end-to-end GPU test (`gpu_olmo3_tokenizer_e2e`)

### HuggingFace Upload
Tokenizer is now unblocked. The model can be correctly encoded/decoded through ATLAS for the `openhubresearch/ATLAS-OLMo-3-7B-Think-v4` upload.

Release: https://github.com/web3guru888/ATLAS/releases/tag/v4.0.4
